### PR TITLE
Translate otlp metrics to prom metatada

### LINKED ISSUE #4910: OTLP Metrics don't send metadata
#### Describe the bug
When ingesting OTLP metrics, we drop the metadata (HELP and TYPE) without propagating it.

#### To Reproduce
Push OTLP Metrics. 

Currently the translation function from upstream OTel Collector doesn't return metadata so we just ignore it: https://github.com/grafana/mimir/blob/4d4a9a9d6bc3a9799e254fe55c7094cf3326d35d/pkg/util/push/otel.go#LL128C1-L128C1

#### Expected behavior

We also append the metadata.

cc @kovrus

### PR DESCRIPTION
**Pull request by jmichalek132**
#### What this PR does

Translate OTLP to Metadata and forward it. 
This makes metrics description and type available when ingest OTLP protocol.

#### Which issue(s) this PR fixes or relates to

Fixes #4910

#### Checklist

- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by gouthamve** [comment]
https://github.com/grafana/mimir/issues/4910#issuecomment-1668452485

1. Enhancement is correct.
2. Yes, but lets do it in a different PR?
3. Going to defer to the maintainers here. But I _think_ so.
4. You can extend it.

**Comment by jmichalek132** [comment]
* Updated tests, please let me know if this is good enough
* Added TODO for exposing addMetricSuffixes to the configuration used [here](https://github.com/grafana/mimir/pull/5693/files#diff-eb47b1b5619b45185602515d8d5274a06451058d43aaa9c85c9fc1a7222635c3R176)
* Waiting for the decision on  feature flag/configuration option to enable this

**Comment by jmichalek132** [comment]
Hi, @gouthamve  @charleskorn could I please bother you for a review?

File: CHANGELOG.md [hunk 1]
```diff
@@ -66,6 +66,7 @@
 * [ENHANCEMENT] Ingester: optimize label-values with matchers call when number of matched series is small. #5600
 * [ENHANCEMENT] Compactor: Delete bucket-index, markers and debug files if there are no blocks left in the bucket index. This cleanup must be enabled by using `-compactor.no-blocks-file-cleanup-enabled` option. #5648
 * [ENHANCEMENT] Ingester: reduce memory usage of active series tracker. #5665
+* [ENHANCEMENT] Distributor: Translate OTLP to Metadata and forward it. This makes metrics description and type available when ingest OTLP protocol. #5693
```
Comment (charleskorn) [inline CHANGELOG.md:69]:
```suggestion
* [ENHANCEMENT] Distributor: store metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingest OTLP protocol. #5693
```

File: integration/e2emimir/client.go [hunk 2]
```diff
@@ -288,6 +290,29 @@ func (c *Client) LabelValuesCardinality(labelNames []string, selector string, li
 	return c.httpClient.Do(req.WithContext(ctx))
 }
 
+// GetPrometheusMetadata fetches the metadata from the Prometheus endpoint /api/v1/metadata.
+func (c *Client) GetPrometheusMetadata(metric string, limit int) (*http.Response, error) {
+	body := make(url.Values)
```
Comment (charleskorn) [inline integration/e2emimir/client.go:295]: Why not just do `url.Values{}` here?

File: integration/e2emimir/client.go [hunk 3]
```diff
@@ -288,6 +290,29 @@ func (c *Client) LabelValuesCardinality(labelNames []string, selector string, li
...
+	if len(metric) > 0 {
+		body.Set("metric", metric)
+	}
+	if limit > 0 {
+		body.Set("limit", strconv.Itoa(limit))
+	}
+
+	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("http://%s/prometheus/api/v1/metadata", c.querierAddress), strings.NewReader(body.Encode()))
```
Comment (charleskorn) [inline integration/e2emimir/client.go:303]: Don't the query parameters (`body`) need to be included in the URL, not the request body?

File: integration/e2emimir/client.go [hunk 4]
```diff
@@ -288,6 +290,29 @@ func (c *Client) LabelValuesCardinality(labelNames []string, selector string, li
...
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("X-Scope-OrgID", c.orgID)
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	defer cancel()
+
+	// Execute HTTP request
```
Comment (charleskorn) [inline integration/e2emimir/client.go:312]: I don't think this comment is necessary.

File: integration/otlp_ingestion_test.go [hunk 5]
```diff
@@ -71,9 +79,34 @@ func TestOTLPIngestion(t *testing.T) {
...
+	   }
+	}
+	`
+
+	require.JSONEq(t, expectedJSON, string(metadataResponseBody))
+
 	// Push series with histograms to Mimir
 	series, expectedVector, _ = generateHistogramSeries("series", now)
-	res, err = c.PushOTLP(series)
+	res, err = c.PushOTLP(series, metadata)
```
Comment (charleskorn) [inline integration/otlp_ingestion_test.go:109]: Might be worth adding a check that the metadata is stored correctly for this histogram series too.

File: pkg/util/push/otel.go [hunk 6]
```diff
@@ -150,10 +152,40 @@ func OTLPHandler(
...
+		scopeMetricsSlice := resourceMetrics.ScopeMetrics()
+
+		for j := 0; j < scopeMetricsSlice.Len(); j++ {
+			scopeMetrics := scopeMetricsSlice.At(j)
+			for k := 0; k < scopeMetrics.Metrics().Len(); k++ {
+				entry := mimirpb.MetricMetadata{
+					Type:             mimirpb.MetricMetadata_MetricType(scopeMetrics.Metrics().At(k).Type()),
+					MetricFamilyName: prometheustranslator.BuildCompliantName(scopeMetrics.Metrics().At(k), "", true), // TODO expose addMetricSuffixes in configuration
+					Help:             scopeMetrics.Metrics().At(k).Description(),
+					Unit:             scopeMetrics.Metrics().At(k).Unit(),
```
Comment (charleskorn) [inline pkg/util/push/otel.go:178]: I'd suggest extracting `scopeMetrics.Metrics().At(k)` out to a variable to reduce the duplication here.

File: pkg/util/push/otel.go [hunk 7]
```diff
@@ -150,10 +152,40 @@ func OTLPHandler(
...
+	for i := 0; i < resourceMetricsSlice.Len(); i++ {
+		resourceMetrics := resourceMetricsSlice.At(i)
+		scopeMetricsSlice := resourceMetrics.ScopeMetrics()
+
+		for j := 0; j < scopeMetricsSlice.Len(); j++ {
+			scopeMetrics := scopeMetricsSlice.At(j)
+			for k := 0; k < scopeMetrics.Metrics().Len(); k++ {
+				entry := mimirpb.MetricMetadata{
+					Type:             mimirpb.MetricMetadata_MetricType(scopeMetrics.Metrics().At(k).Type()),
+					MetricFamilyName: prometheustranslator.BuildCompliantName(scopeMetrics.Metrics().At(k), "", true), // TODO expose addMetricSuffixes in configuration
```
Comment (charleskorn) [inline pkg/util/push/otel.go:176]: Is this TODO still relevant?

**Review by charleskorn — COMMENTED** [review]
Thanks for putting this together @jmichalek132. 

As you suggested in [your comment on the issue](https://github.com/grafana/mimir/issues/4910#issuecomment-1668452485), I think it would be good to add a configuration flag to allow enabling / disabling this. For now, let's mark it as experimental and default to disabled until we've had a chance to test this at scale.

I'd also like to get a review from someone more familiar with OTLP before we merge, as I'm not an OTLP expert.

File: pkg/util/push/otel.go (hunk 6 above)
Comment (jmichalek132) [reply pkg/util/push/otel.go]: Done.

File: integration/e2emimir/client.go (hunk 4 above)
Comment (jmichalek132) [reply integration/e2emimir/client.go]: Removed

**Comment by jmichalek132** [comment]
Is there a way to run the linter that will also apply the formatting?

File: integration/e2emimir/client.go (hunk 2 above)
Comment (jmichalek132) [reply integration/e2emimir/client.go]: Removed

File: integration/otlp_ingestion_test.go (hunk 5 above)
Comment (jmichalek132) [reply integration/otlp_ingestion_test.go:109]: Added.

File: pkg/util/push/otel.go (hunk 7 above)
Comment (jmichalek132) [reply pkg/util/push/otel.go]: Yes as per @gouthamve request, exposing this boolean into configuration should be done in separate PR
https://github.com/grafana/mimir/pull/5693#issuecomment-1669216300
Answering my question
https://github.com/grafana/mimir/issues/4910#issuecomment-1668452485

File: pkg/util/push/otel.go [hunk 8]
```diff
@@ -45,6 +45,7 @@ func OTLPHandler(
 	maxRecvMsgSize int,
 	sourceIPs *middleware.SourceIPExtractor,
 	allowSkipLabelNameValidation bool,
+	EnableOtelMetadataStorage bool,
```
Comment (charleskorn) [inline pkg/util/push/otel.go:48]: The convention in Go is for parameter names to be camelCase.

```suggestion
	enableOtelMetadataStorage bool,
```

File: pkg/util/push/push_test.go
Comment (charleskorn) [inline pkg/util/push/push_test.go:1]: Could you please add a test for the case where the feature flag is disabled?

File: pkg/api/api.go [hunk 9]
```diff
@@ -50,6 +50,9 @@ type ConfigHandler func(actualCfg interface{}, defaultCfg interface{}) http.Hand
 type Config struct {
 	SkipLabelNameValidationHeader bool `yaml:"skip_label_name_validation_header_enabled" category:"advanced"`
 
+	// when true the distributor translate OTLP to metadata and forwards it,
```
Comment (charleskorn) [inline pkg/api/api.go:53]: I don't think this comment is necessary, we have the description on the flag to help people understand what this is for.

File: pkg/api/api.go [hunk 10]
```diff
@@ -67,6 +70,7 @@ type Config struct {
 // RegisterFlags adds the flags required to config this to the given FlagSet.
 func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.SkipLabelNameValidationHeader, "api.skip-label-name-validation-header-enabled", false, "Allows to skip label name validation via X-Mimir-SkipLabelNameValidation header on the http write path. Use with caution as it breaks PromQL. Allowing this for external clients allows any client to send invalid label names. After enabling it, requests with a specific HTTP header set to true will not have label names validated.")
+	f.BoolVar(&cfg.EnableOtelMetadataStorage, "distributor.enable-otlp-metadata-storage", false, "Distributor: store metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingest OTLP protocol.")
```
Comment (charleskorn) [inline pkg/api/api.go:73]: [nit]

```suggestion
	f.BoolVar(&cfg.EnableOtelMetadataStorage, "distributor.enable-otlp-metadata-storage", false, "If true, store metadata when ingesting metrics via OTLP. This makes metric descriptions and types available for metrics ingested via OTLP.")
```

File: pkg/util/push/otel.go [hunk 11]
```diff
@@ -150,10 +153,43 @@ func OTLPHandler(
...
+			metadata := otelMetricsToMetadata(otlpReq.Metrics())
+			req.Metadata = metadata
+		}
+
 		return body, nil
 	})
 }
 
+func otelMetricsToMetadata(md pmetric.Metrics) []*mimirpb.MetricMetadata {
+	var metadata []*mimirpb.MetricMetadata
```
Comment (charleskorn) [inline pkg/util/push/otel.go:167]: One small performance improvement here would be to pre-allocate a slice large enough to hold all of the expected metadata for this request.

**Comment by charleskorn** [comment]
> Is there a way to run the linter that will also apply the formatting?

`golangci-lint run --fix ./...` will automatically fix some issues.

File: pkg/util/push/otel.go (hunk 8 above)
Comment (jmichalek132) [reply pkg/util/push/otel.go]: fixed

File: pkg/api/api.go (hunk 9 above)
Comment (jmichalek132) [reply pkg/api/api.go]: removed

**Comment by colega** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

File: pkg/util/push/otel.go (hunk 11 above)
Comment (jmichalek132) [reply pkg/util/push/otel.go]: updated

File: pkg/api/api.go [hunk 12]
```diff
@@ -50,6 +50,8 @@ type ConfigHandler func(actualCfg interface{}, defaultCfg interface{}) http.Hand
 type Config struct {
 	SkipLabelNameValidationHeader bool `yaml:"skip_label_name_validation_header_enabled" category:"advanced"`
 
+	enableOtelMetadataStorage bool `yaml:"enable_otel_metadata_translation" category:"experimental"`
```
Comment (charleskorn) [inline pkg/api/api.go:53]: Could you please add this to the list of experimental features in [the docs](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md)?

File: pkg/util/push/otel.go [hunk 13]
```diff
@@ -150,10 +152,49 @@ func OTLPHandler(
...
+		resourceMetrics := resourceMetricsSlice.At(i)
+		scopeMetricsSlice := resourceMetrics.ScopeMetrics()
+
+		for j := 0; j < scopeMetricsSlice.Len(); j++ {
+			scopeMetrics := scopeMetricsSlice.At(j)
+			for k := 0; k < scopeMetrics.Metrics().Len(); k++ {
+				metric := scopeMetrics.Metrics().At(k)
+				entry := mimirpb.MetricMetadata{
+					Type:             mimirpb.MetricMetadata_MetricType(metric.Type()),
+					MetricFamilyName: prometheustranslator.BuildCompliantName(metric, "", true), // TODO expose addMetricSuffixes in configuration
```
Comment (charleskorn) [inline pkg/util/push/otel.go:184]: Is this TODO something you're planning to address in this PR?

Comment (jmichalek132) [reply pkg/util/push/otel.go:184]: No, @gouthamve asked to leave the implementation of that for follow up PR here https://github.com/grafana/mimir/pull/5693#issuecomment-1669216300, should I remove the comment and create an issue for it?

File: pkg/api/api.go (hunk 12 above)
Comment (jmichalek132) [reply pkg/api/api.go:53]: will do

File: pkg/util/push/otel.go (hunk 13 above)
Comment (gouthamve) [reply pkg/util/push/otel.go:184]: I'd create an issue yup.

Comment (jmichalek132) [reply pkg/util/push/otel.go:184]: Created https://github.com/grafana/mimir/issues/5967. Should I remove the comment or leave it in?

File: pkg/api/api.go (hunk 12 above)
Comment (jmichalek132) [reply pkg/api/api.go:53]: done, is it ok?

File: pkg/util/push/otel.go (hunk 13 above)
Comment (gouthamve) [reply pkg/util/push/otel.go:184]: I'd put #5967 in the comment tbh. I'd leave it in.

File: pkg/util/push/push_test.go
Comment (jmichalek132) [reply pkg/util/push/push_test.go:1]: Added here https://github.com/grafana/mimir/pull/5693/files#diff-9cfabdded42ea41657229be85019e9543e391f591bc1aaee09fd36d8b83fc218R130 is this okay?

File: CHANGELOG.md [hunk 14]
```diff
@@ -8,6 +8,7 @@
 * [ENHANCEMENT] Query-frontend: add `cortex_query_frontend_enqueue_duration_seconds` metric that records the time taken to enqueue or reject a query request when not using the query-scheduler. #5879
 * [ENHANCEMENT] Expose `/sync/mutex/wait/total:seconds` Go runtime metric as `go_sync_mutex_wait_total_seconds_total` from all components. #5879
 * [ENHANCEMENT] Go: updated to 1.21.1. #5955
+* [ENHANCEMENT] Distributor: store metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingest OTLP protocol. #5693
```
Comment (charleskorn) [inline CHANGELOG.md:11]:
```suggestion
* [ENHANCEMENT] Distributor: store metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingesting metrics via OTLP. #5693
```

**Review by charleskorn — APPROVED** [review]
LGTM, thanks @jmichalek132!

File: CHANGELOG.md (hunk 14 above)
Comment (charleskorn) [inline CHANGELOG.md]:
```suggestion
* [ENHANCEMENT] Distributor: add experimental support for storing metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingest OTLP protocol. Enable with `-distributor.enable-otlp-metadata-storage=true`. #5693
```

File: CHANGELOG.md [hunk 15]
```diff
@@ -8,6 +8,7 @@
 * [ENHANCEMENT] Query-frontend: add `cortex_query_frontend_enqueue_duration_seconds` metric that records the time taken to enqueue or reject a query request when not using the query-scheduler. #5879
 * [ENHANCEMENT] Expose `/sync/mutex/wait/total:seconds` Go runtime metric as `go_sync_mutex_wait_total_seconds_total` from all components. #5879
 * [ENHANCEMENT] Go: updated to 1.21.1. #5955
+* [ENHANCEMENT] Distributor: store metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingesting metrics via OTLP. #5693
```
Comment (charleskorn) [inline CHANGELOG.md:11]:
```suggestion
* [ENHANCEMENT] Distributor: add experimental support for storing metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingesting metrics via OTLP. Enable with `-distributor.enable-otlp-metadata-storage=true`. #5693
```

**Comment by charleskorn** [comment]
Could you please fix the linting and test failures @jmichalek132?

**Comment by jmichalek132** [comment]
> Could you please fix the linting and test failures @jmichalek132?

All should be fixed now. Can you please approve the test run again?

**Review by charleskorn — APPROVED** [review]

**Comment by charleskorn** [comment]
@gouthamve would you mind taking another look at this when you get a chance?

File: pkg/util/push/otel.go [hunk 16]
```diff
@@ -150,10 +152,49 @@ func OTLPHandler(
...
+	for i := 0; i < resourceMetricsSlice.Len(); i++ {
+		resourceMetrics := resourceMetricsSlice.At(i)
+		scopeMetricsSlice := resourceMetrics.ScopeMetrics()
+
+		for j := 0; j < scopeMetricsSlice.Len(); j++ {
+			scopeMetrics := scopeMetricsSlice.At(j)
+			for k := 0; k < scopeMetrics.Metrics().Len(); k++ {
+				metric := scopeMetrics.Metrics().At(k)
+				entry := mimirpb.MetricMetadata{
+					Type:             mimirpb.MetricMetadata_MetricType(metric.Type()),
```
Comment (gouthamve) [inline pkg/util/push/otel.go:183]: I'm confused here. `metric.Type()` returns this: https://pkg.go.dev/go.opentelemetry.io/collector/pdata/pmetric#MetricType

While https://github.com/grafana/mimir/blob/main/pkg/mimirpb/mimir.pb.go#L53-L64 is a different thing. Specifically OTLP has "Sum" while Mimir has "COUNTER". How are we mapping them together?

Comment (jmichalek132) [reply pkg/util/push/otel.go:183]: Yeah I will look into how is this handled in https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusexporter

Comment (jmichalek132) [reply pkg/util/push/otel.go]: Added the function for translating OTEL metric types to Prometheus metric types.

File: pkg/util/push/otel.go [hunk 17]
```diff
@@ -150,10 +152,69 @@ func OTLPHandler(
...
+		}
+		return metricType
+	case pmetric.MetricTypeHistogram:
+		return mimirpb.MetricMetadata_MetricType(3)
+	case pmetric.MetricTypeSummary:
+		return mimirpb.MetricMetadata_MetricType(5)
+	case pmetric.MetricTypeExponentialHistogram:
+		return mimirpb.MetricMetadata_MetricType(3)
+	}
+	return mimirpb.MetricMetadata_MetricType(0)
```
Comment (jmichalek132) [inline pkg/util/push/otel.go:182]: Should I add warning / debug log message here when the default case happens?

File: pkg/util/push/otel.go [hunk 18]
```diff
@@ -162,6 +162,26 @@ func OTLPHandler(
...
```
Comment (gouthamve) [inline pkg/util/push/otel.go:182]:
```suggestion
	case pmetric.MetricTypeGauge:
		return mimirpb.GAUGE
	case pmetric.MetricTypeSum:
		metricType := mimirpb.GAUGE
		if otelMetric.Sum().IsMonotonic() {
			metricType = mimirpb.COUNTER
		}
		return metricType
	case pmetric.MetricTypeHistogram:
		return mimirpb.HISTOGRAM
	case pmetric.MetricTypeSummary:
		return mimirpb.SUMMARY
	case pmetric.MetricTypeExponentialHistogram:
		return mimirpb.HISTOGRAM
	}
	return mimirpb.UNKNOWN
```

File: pkg/util/push/push_test.go [hunk 19]
```diff
@@ -74,7 +74,7 @@ func TestHandlerOTLPPush(t *testing.T) {
 		assert.Equal(t, "foo", series[0].Labels[0].Value)
 
 		metadata := request.Metadata
-		assert.Equal(t, mimirpb.MetricMetadata_MetricType(1), metadata[0].GetType())
```
Comment (gouthamve) [inline pkg/util/push/push_test.go]: Similar changes here.

Comment (jmichalek132) [reply pkg/util/push/push_test.go]: fixed

**Review by gouthamve — APPROVED** [review]
LGTM! @charleskorn This is ready to go imo!

**Pull request merged**